Tests/TABResourceLoaderTests/Protocols/NetworkResourceTypeTests.swift

```diff
@@ -79,6 +81,51 @@ class NetworkResourceTypeTests: XCTestCase {
     XCTAssertEqual(urlRequest?.url?.absoluteString, urlWithQueryParameters.absoluteString)
   }
 
+  func test_urlRequest_resourceURLWithQueryParametersRemovingAllowedCharacterPlusAndColon() {
+    let mockedURLQueryItems = [URLQueryItem(name: "query:with+another-name", value: "query:value+another-value")]
+
+    var allowedCharacterSet: CharacterSet = .urlQueryAllowed
+    allowedCharacterSet.remove("+")
+    allowedCharacterSet.remove(":")
+
+    let resource = MockCustomNetworkResource(url: urlWithQueryItem, queryItems: mockedURLQueryItems, urlQueryAllowedCharacterSet: allowedCharacterSet)
+
+    let expectedURLString = "\(urlWithQueryItem)&query%3Awith%2Banother-name=query%3Avalue%2Banother-value"
+    let urlRequest = resource.urlRequest()
+
+    XCTAssertEqual(urlRequest?.url?.absoluteString, expectedURLString)
+  }
+
+  func test_urlRequest_resourceURLWithQueryParametersUsingAllowedCharacterPlusAndColon() {
+    let mockedURLQueryItems = [URLQueryItem(name: "query:with+another-name", value: "query:value+another-value")]
+    let resource = MockCustomNetworkResource(url: urlWithQueryItem, queryItems: mockedURLQueryItems)
+
+    let expectedURLString = "\(urlWithQueryItem)&query:with+another-name=query:value+another-value"
+    let urlRequest = resource.urlRequest()
+
+    XCTAssertEqual(urlRequest?.url?.absoluteString, expectedURLString)
+  }
+
+  func test_urlRequest_resourceURLWithQueryParametersWithEmptyParameter() {
+    let mockedURLQueryItems = [URLQueryItem(name: "query-name", value: "")]
+    let resource = MockCustomNetworkResource(url: urlWithQueryItem, queryItems: mockedURLQueryItems)
+
+    let expectedURLString = "\(urlWithQueryItem)&query-name="
+    let urlRequest = resource.urlRequest()
+
+    XCTAssertEqual(urlRequest?.url?.absoluteString, expectedURLString)
+  }
+
+  func test_urlRequest_resourceURLWithQueryParametersWithMalformedParameter() {
+    let mockedURLQueryItems = [URLQueryItem(name: "query-name", value: "malformed&value")]
+    let resource = MockCustomNetworkResource(url: urlWithQueryItem, queryItems: mockedURLQueryItems)
+
+    let expectedURLString = "\(urlWithQueryItem)&query-name=malformed&value"
```

Comment:
Sorry, I shouldn't have said "malformed". What I was trying to say is what if the value has an ampersand or colon in it? Its not illegal to have them as values. They should get encoded to `%26` & `%3A` respectively.

For example, this is ok - `http://foo.bar/search?brand=Marks&Spencer`. The `brand` query param will get encoded to `brand=Marks%26Spender`.

I wonder if we need the default character set which does not contain those characters. Maybe we should just provide `.urlQueryAllowed` set as default.

Reply:
With .urlQueryAllowed the URLQueryItem parser will not encode & and =, so I needed to put the CharacterSet .urlQueryAllowed removing those characters to let the parser encode them.
I just forgot to set .improvedUrlQueryAllowed as the allowedCharacterSet of the test class!

Reply:
Also changed the name of the function to be more readable

Reply:
Great!